Add solutions and update functional programming challenge instructions

petern-sc opened the pull request:
Added solutions to the functional programming exercises I've touched, and made some minor changes to them. 

- Added solutions to some of the guides, in both vanilla and ES6. I think it would be good to show both approaches, as these examples can be a good reference point for people. 
- Changed the `avoid-mutations-and-side-effects...` exercise to use a passed in argument instead of the global variable. This is more in line with the idea of using functions that provide outputs which are purely dependent on their inputs. 
- Edited the `use-the-map-method...` exercise to stringify the console output, so people doing the exercise can actually see their outputs. 

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] My article does not contain shortened URLs or affiliate links.

Commit:
feat: Update intro to currying

- Add solutions
- Add ES6 examples

Commit:
feat: Add solution to function-programming map

- Add solutions, both vanilla and ES6 arrow functions
- Add JSON.stringify to console.log, makes it easier to see what the resulting output is rather than showing
e.g. [object, object, object]

RandellDawson reviewed (changes requested):
@petern-sc I have reviewed your commits, but have a few suggestions before making the final approval.  

For the avoid-mutations-and-side-effects..., it is best to leave the current solution seed, because the next challenge after this one deals with passing the arguments into the function instead of referencing the global variable.  Making the change you suggested would create almost a duplicate challenge.

For the introduction-to-currying-and-partial-application.english, everything looks fine.

For the use-the-map-method-to-extract-data-from-an-array, the solution only needs a single solution, because this is not part of the Guide.  Both solutions are fine, so just pick your favorite.

Let me know when you have made the changes.

Thank you for your request.

Commit:
feat: Add solution to FP mutations

RandellDawson commented:
@petern-sc It appears you still have two solutions for the Avoid Mutations and Side Effects Using Functional Programming challenge.  It only needs a single solution.  Just go with the ES6 version.

RandellDawson committed:
update: used the contributor's ES6 solution

Contributor originally had a non-ES6 solution and an ES6 solution.  I removed the non-ES6 solution.
